### PR TITLE
Broker Strategies and the protocol verification with MSAL | part2/3

### PR DESCRIPTION
**Pull request by heidijinxujia**
This change is limited to multiAccountsPCA.

with compatible broker
- Broker Version 3.0
- MSAL Client required version 3.0

|  with compatible broker version      | MicrosoftAuthService           | AccountManager  |
| ------------- |:-------------:| -----:|
| AT    | passed | passed|
| ATS    | passed      |   fixed* |
| GetAccounts | passed     |    fixed* |
| RemoveAccount | passed    |    fixed* |
* this is an unrelated bug, filed #682 

with incompatible broker
- Broker Version 2.0
- MSAL Client required version 3.0

|   with incompatible broker version     | MicrosoftAuthService           | AccountManager  |
| ------------- |:-------------:| -----:|
| AT    | passed | passed|
| ATS    |passed      |  passed |
| GetAccounts | passed     |    passed |
| RemoveAccount | passed      |    passed|

### CODE REVIEW

File: msal/src/main/java/com/microsoft/identity/client/internal/controllers/BrokerMsalController.java [hunk 1]
```diff
@@ -1074,6 +1082,114 @@ private Bundle getRequestBundleForRemoveAccountFromSharedDevice(PublicClientAppl
         return requestBundle;
     }
 
+    private static boolean helloWithMicrosoftAuthService(@NonNull final OperationParameters parameters) throws ClientException {
```
Comment (shoatman) [inline msal/src/main/java/com/microsoft/identity/client/internal/controllers/BrokerMsalController.java]: Should we add the worker thread annotation here as well?

File: testapps/testapp/src/main/res/raw/msal_config.json [hunk 2]
```diff
@@ -4,6 +4,7 @@
   "redirect_uri" : "msauth://com.microsoft.identity.client.sample.local/1wIqXSqBj7w%2Bh11ZifsnqwgyKrY%3D",
   "multiple_clouds_supported":true,
   "broker_redirect_uri_registered": true,
+  "minimum_required_broker_protocol_version":"3.0",
```
Comment (heidijinxujia) [inline testapps/testapp/src/main/res/raw/msal_config.json]: move this into default config.

File: msal/src/main/java/com/microsoft/identity/client/internal/controllers/BrokerAuthServiceStrategy.java [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+        final MsalBrokerRequestAdapter msalBrokerRequestAdapter = new MsalBrokerRequestAdapter();
+        interactiveRequestIntent.putExtra(
+                AuthenticationConstants.Broker.BROKER_REQUEST_V2,
+                new Gson().toJson(
+                        msalBrokerRequestAdapter.brokerRequestFromAcquireTokenParameters(parameters),
+                        BrokerRequest.class)
+        );
+        interactiveRequestIntent.putExtra(AuthenticationConstants.Broker.ACCOUNT_NAME, parameters.getLoginHint());
+
+
```
Comment (heidijinxujia) [inline msal/src/main/java/com/microsoft/identity/client/internal/controllers/BrokerAuthServiceStrategy.java]: extra line

File: msal/src/main/java/com/microsoft/identity/client/internal/controllers/BrokerMsalController.java [hunk 4]
```diff
@@ -1111,4 +768,31 @@ private static boolean isPermissionGranted(@NonNull final Context context,
         Logger.verbose(TAG + methodName, "is " + permissionName + " granted? [" + isGranted + "]");
         return isGranted;
     }
-}
+
+    private void initializeBrokerMsalController(@NonNull final OperationParameters parameters)
```
Comment (heidijinxujia) [inline msal/src/main/java/com/microsoft/identity/client/internal/controllers/BrokerMsalController.java]: rename the function to reflect "sayAloha"

File: msal/src/main/java/com/microsoft/identity/client/internal/controllers/BrokerMsalController.java [hunk 5]
```diff
@@ -1111,4 +768,31 @@ private static boolean isPermissionGranted(@NonNull final Context context,
...
+
+        //check if bound service available
+        if (BrokerMsalController.helloWithMicrosoftAuthService(parameters.getAppContext(), parameters)) {
+            Logger.verbose(TAG + methodName, "Add the broker AuthService strategy.");
+            this.addBrokerStrategy(new BrokerAuthServiceStrategy());
+        }
+
+        //check if account manager available
+        if (BrokerMsalController.helloWithAccountManager(parameters.getAppContext(), parameters)) {
+            Logger.verbose(TAG + methodName, "Add the account manager strategy.");
```
Comment (heidijinxujia) [inline msal/src/main/java/com/microsoft/identity/client/internal/controllers/BrokerMsalController.java]: Need to check the permission here.

Comment (heidijinxujia) [reply msal/src/main/java/com/microsoft/identity/client/internal/controllers/BrokerMsalController.java]: Addressed.

**Pull request merged**